code/parse/sexp.cpp

```diff
@@ -23298,6 +23567,706 @@ void sexp_set_motion_debris(int node)
+			return SEXP_FALSE;
+		}
+
+		node = CDR(node);
+	}
+
+	return SEXP_TRUE;
+}
+
+/**
+*	SEXP which returns the index in a SEXP list container for a specific element.
+*	Returns -1 if the element is not found.
+*/
+int sexp_list_data_index(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return -1;
+	}
+
+	if (!Sexp_containers[index].is_list()) {
+		Warning(LOCATION, "List-data-index called on non-list container %s.", container_name);
+		return -1;
+	}
+
+	node = CDR(node);
+	Assert(node != -1);
+
+	const SCP_string possible_data = CTEXT(node);
+
+	int data_index = 0;
+	for (const SCP_string &data : Sexp_containers[index].list_data) {
+		if (possible_data == data) {
+			return data_index;
+		}
+		++data_index;
+	}
+
+	return -1;
+}
+
+/**
+* Check if a SEXP map container contains specific keys.
+*/
+int sexp_map_has_key(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return SEXP_FALSE;
+	}
+
+	if (!Sexp_containers[index].is_map()) {
+		Warning(LOCATION, "Map-has-key called on non-map container %s.", container_name);
+		return SEXP_FALSE;
+	}
+
+	const auto &map_data = Sexp_containers[index].map_data;
+
+	node = CDR(node);
+	// there should be at least one key to search for
+	Assert(node != -1);
+
+	SCP_string possible_key;
+	while (node != -1) {
+		possible_key = CTEXT(node);
+		if (map_data.find(possible_key) == map_data.end()) {
+			return SEXP_FALSE;
+		}
+
+		node = CDR(node);
+	}
+
+	return SEXP_TRUE;
+}
+
+/**
+* Check if a SEXP map container has a key whose data matches a specific string.
+* If a third argument (string variable) is provided, store the data's associated key, if a match is found.
+* Returns -1 if the element is not in the container
+*/
+int sexp_map_has_data_item(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return SEXP_FALSE;
+	}
+
+	if (!Sexp_containers[index].is_map()) {
+		Warning(LOCATION, "map-has-data-teim called on non-map container %s.", container_name);
+		return SEXP_FALSE;
+	}
+
+	node = CDR(node);
+	Assert(node != -1);
+
+	const SCP_string possible_data = CTEXT(node);
+
+	for (const auto &kv_pair : Sexp_containers[index].map_data) {
+		if (possible_data == kv_pair.second) {
+			// check for optional variable to store the key
+			node = CDR(node);
+			if (node != -1) {
+				const int var_index = sexp_get_variable_index(node);
+				// DISCUSSME: should we do an explicit if-check instead?
+				Assert(var_index >= 0);
+				// TODO: type checking of string key/var vs. number key/var?
```

Comment:
Should the check that the variable is the same type  as the map's keys (number or string) be done in `check_sexp_syntax()`? Can we assume that, if we get to this point, the types have already been checked?